httpbp/client_middlewares.go

```diff
@@ -0,0 +1,225 @@
+package httpbp
+
+import (
+	"errors"
+	"io"
+	"net/http"
+	"sync"
+	"sync/atomic"
+
+	retry "github.com/avast/retry-go"
+	"github.com/opentracing/opentracing-go"
+
+	"github.com/reddit/baseplate.go/breakerbp"
+	"github.com/reddit/baseplate.go/retrybp"
+	"github.com/reddit/baseplate.go/tracing"
+)
+
+const (
+	defaultMaxErrorReadAhead = 1024
+)
+
+// ClientMiddleware is used to build HTTP client middleware by implementing
+// http.RoundTripper which http.Client accepts as Transport.
+type ClientMiddleware func(next http.RoundTripper) http.RoundTripper
+
+// roundTripperFunc adapts closures and functions to implement http.RoundTripper.
+type roundTripperFunc func(req *http.Request) (*http.Response, error)
+
+func (f roundTripperFunc) RoundTrip(req *http.Request) (*http.Response, error) {
+	return f(req)
+}
+
+// NewClient returns a standard HTTP client wrapped with the default middleware
+// plus any additional client middleware passed into this function. Default
+// middlewares are: MonitorClient and Retries. ClientErrorWrapper is included
+// as transitive middleware through Retries.
+func NewClient(config ClientConfig, middleware ...ClientMiddleware) *http.Client {
+	transport := &http.Transport{
+		MaxConnsPerHost: config.MaxConnections,
+	}
+
+	defaults := []ClientMiddleware{
+		MonitorClient(config.Slug),
+		Retries(
+			config.MaxErrorReadAhead,
+			retry.Attempts(1),
+			retrybp.Filters(
+				retrybp.NetworkErrorFilter,
+				retrybp.RetryableErrorFilter,
+			),
```

Comment:
1. `RetryableErrorFilter` should be the first in the filter chain, see [its documentation](https://pkg.go.dev/github.com/reddit/baseplate.go/retrybp#RetryableErrorFilter) for why
2. We shouldn't be adding a default, hardcoded filter chain. This should be part of `ClientConfig`, and if it's unset (default value), we should only add `retry.Attempts(1)` (so users can control per request retry behavior via context object).

Reply:
For 2: do you think it's sufficient to have a if `retries > 1` then apply the filters? Or have the config take an `[]retry.Option` similar to how we do it for Thrift